tasks/seal/task_submit_commit.go

```diff
@@ -158,170 +146,345 @@ func (s *SubmitCommitTask) Do(taskID harmonytask.TaskID, stillOwned func() bool)
+	}
+
+	balance, err := s.api.StateMinerAvailableBalance(ctx, maddr, types.EmptyTSK)
+	if err != nil {
+		if err != nil {
+			return nil, xerrors.Errorf("getting miner balance: %w", err)
 		}
-		balance, err := s.api.StateMinerAvailableBalance(ctx, maddr, types.EmptyTSK)
+	}
+
+	mi, err := s.api.StateMinerInfo(ctx, maddr, types.EmptyTSK)
+	if err != nil {
+		return nil, xerrors.Errorf("getting miner info: %w", err)
+	}
+
+	if len(infos) > 4 {
+		arp, err := aggregateProofType(nv)
 		if err != nil {
-			if err != nil {
-				return false, xerrors.Errorf("getting miner balance: %w", err)
-			}
+			return nil, xerrors.Errorf("getting aggregate proof type: %w", err)
 		}
-		collateral = big.Sub(collateral, balance)
-		if collateral.LessThan(big.Zero()) {
-			collateral = big.Zero()
+		aggParams.AggregateProofType = &arp
+
+		aggParams.AggregateProof, err = s.prover.AggregateSealProofs(proof.AggregateSealVerifyProofAndInfos{
+			Miner:          abi.ActorID(SpID),
+			SealProof:      sealProof,
+			AggregateProof: arp,
+			Infos:          infos,
+		}, aggParams.SectorProofs)
+
+		if err != nil {
+			return nil, xerrors.Errorf("aggregating proofs: %w", err)
+		}
+		aggParams.SectorProofs = nil // can't be set when aggregating
+
+		aggFeeRaw, err := policy.AggregateProveCommitNetworkFee(nv, len(infos), ts.MinTicketBlock().ParentBaseFee)
+		if err != nil {
+			return nil, xerrors.Errorf("getting aggregate commit network fee: %s", err)
 		}
+
+		aggFee := big.Div(big.Mul(aggFeeRaw, big.NewInt(110)), big.NewInt(110))
+
+		aggCollateral := big.Add(collateral, aggFee)
+		aggCollateral = s.calculateCollateral(balance, aggCollateral)
+		goodFunds := big.Add(maxFee, aggCollateral)
+		enc := new(bytes.Buffer)
+		if err := params.MarshalCBOR(enc); err != nil {
+			return nil, xerrors.Errorf("could not serialize commit params: %w", err)
+		}
+		aggMsg, err = s.gasEstimateCommit(ctx, maddr, enc.Bytes(), mi, goodFunds, aggCollateral, maxFee, ts.Key())
+		if err != nil {
+			return nil, err
+		}
+		aggGas := big.Mul(big.Add(ts.MinTicketBlock().ParentBaseFee, aggMsg.GasPremium), big.NewInt(aggMsg.GasLimit))
+		aggCost = big.Add(aggGas, aggFee)
 	}
 
-	a, _, err := s.as.AddressFor(ctx, s.api, maddr, mi, api.CommitAddr, collateral, big.Zero())
+	{
+		collateral = s.calculateCollateral(balance, collateral)
+		goodFunds := big.Add(maxFee, collateral)
+		enc := new(bytes.Buffer)
+		if err := params.MarshalCBOR(enc); err != nil {
+			return nil, xerrors.Errorf("could not serialize commit params: %w", err)
+		}
+		msg, err = s.gasEstimateCommit(ctx, maddr, enc.Bytes(), mi, goodFunds, collateral, maxFee, ts.Key())
+		if err != nil {
+			return nil, err
+		}
+		gas := big.Mul(big.Add(ts.MinTicketBlock().ParentBaseFee, msg.GasPremium), big.NewInt(msg.GasLimit))
+		cost = gas
+	}
+
+	if aggCost.Nil() || cost.LessThan(aggCost) {
+		log.Debugw("Sending commit message without aggregate", "Batch Cost", cost, "Aggregate Cost", aggCost)
```

Comment:
I'd add a log.infow above this if always printing the decision for now; Can demote to debug once we gain confidence that it works well